File: CPP - old/sorting/mergesort.cpp
//stable O(nlogn)

#include<iostream>
using namespace std;

void merge(int l[],int r[],int a[],int nl,int nr);
void mergesort(int a[],int n)
{
	if(n<2)
	 return;
	int mid=n/2;
	int l[mid];
	int r[n-mid];
	
	for(int i=0;i<mid;i++)
	 l[i]=a[i];
	 
	for(int i=mid,j=0;i<n;i++,j++)
	 r[j]=a[i];
	 
	mergesort(l,mid);
	mergesort(r,n-mid);
	merge(l,r,a,mid,n-mid);
}

void merge(int l[],int r[],int a[],int nl,int nr)
{
	int i=0,j=0,k=0;
	
	while(i<nl && j<nr)
	{
	  if(l[i]<r[j]){
	  	a[k]=l[i];
	  	i++;
	  	k++;
	  }	
	  else{
	  	a[k]=r[j];
	  	j++;
	  	k++;
	  }
	}
	while(i<nl){
		a[k]=l[i];
		k++;
		i++;
	}
	while(j<nr){
		a[k]=r[j];
		k++;
		j++;
	}
}
int main(){
	int a[]={3,4,2,5,1,8,7,6};
	int n=sizeof(a)/sizeof(a[0]);
	for(int i=0;i<n;i++)
	 cout<<a[i]<<" ";
	cout<<endl;
	mergesort(a,n);
	for(int i=0;i<n;i++)
	 cout<<a[i]<<" ";
	return 0;
}

File: Java/Sorting algos/MergeSort.java
import java.util.ArrayList;
import java.util.Arrays;

public class MergeSort {
  public static void main(String[] args){
    ArrayList<Integer> nums = new ArrayList<>(Arrays.asList(30,13,56,15,12,5,7,1));
    sort(nums, 0, nums.size()-1);
    System.out.println("Sorted array : " + nums);
  }

  public static void sort(ArrayList<Integer> nums, int start, int end){
    if(start < end){
      int mid = (start + end)/2;
      sort(nums, start, mid);
      sort(nums, mid+1, end);
      merge(nums, start, mid, end);
    }
  }

  public static void merge(ArrayList<Integer> nums, int start, int mid, int end){

    ArrayList<Integer> leftArr = new ArrayList<>(nums.subList(start, mid+1));
    ArrayList<Integer> rightArr = new ArrayList<>(nums.subList(mid+1, end+1));

    int pointer1 = 0;
    int pointer2 = 0;

    // logic 1

    for(int i= start; i<=end; i++){
      if(pointer1<leftArr.size() && ((pointer2 >= rightArr.size()) || (leftArr.get(pointer1) < rightArr.get(pointer2)))){
        nums.set(i,leftArr.get(pointer1));
        pointer1++;
      }
      else{
        nums.set(i,rightArr.get(pointer2));
        pointer2++;
      }
    }


    // logic 2 (simple)

    // int mainPointer = start;

    // while(pointer1<leftArr.size() && pointer2<rightArr.size()){
    //   if((leftArr.get(pointer1) < rightArr.get(pointer2))){
    //     nums.set(mainPointer,leftArr.get(pointer1));
    //     pointer1++;
    //   }
    //   else{
    //     nums.set(mainPointer,rightArr.get(pointer2));
    //     pointer2++;
    //   }
    //   mainPointer++;
    // }

    // while(pointer1<leftArr.size()) {
    //   nums.set(mainPointer,leftArr.get(pointer1));
    //   pointer1++;
    //   mainPointer++;
    // }

    // while(pointer2<rightArr.size()) {
    //   nums.set(mainPointer,rightArr.get(pointer2));
    //   pointer2++;
    //   mainPointer++;
    // }
  }
}
